jobserver/views/projects.py

```diff
@@ -161,21 +172,25 @@ def get_outputs(self, workspaces):
 
         return Snapshot.objects.filter(pk__in=snapshot_pks).order_by("-published_at")
 
-    def get_repos(self, repo_urls):
-        for url in repo_urls:
+    def iter_repos(self, repo_urls):
+        def get_repo(url):
             f = furl(url)
 
             try:
                 is_private = get_repo_is_private(*f.path.segments)
             except requests.HTTPError:
                 is_private = None
 
-            yield {
+            return {
                 "name": f.path.segments[1],
                 "url": url,
                 "is_private": is_private,
             }
 
+        futures = [repo_thread_pool.submit(get_repo, url=url) for url in repo_urls]
+        for future in concurrent.futures.as_completed(futures):
```

Comment:
So, I think we can slightly simplify this using [Executor.map()](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.map). We can set an explicit timeout for *all* requests too, which is usually a good idea when doing concurrency.

So, maybe replace the last 3 lines with:

```python
# use the threadpool to parallelise the repo requests
yield from repo_thread_pool.map(get_repo, repo_urls, timeout=30)
```

Reply:
Much nicer, and TIL!